### ISSUE TITLE
Add a compiler error note for reference to a previously-failed-to-be-declared identifier

### ISSUE THREAD

**Issue by hryx** [ziglang/zig#2285]
Consider the error messages produced in this example:

```zig
const S = struct {
    value: bool,
};

pub fn main() void {
    // error: use of undeclared identifier 'bogus'
    // This is expected, and happens whether the type is explicit or inferred.
    const totally_a_declared_identifier = S{ .value = bogus };

    // error: use of undeclared identifier 'totally_a_declared_identifier'
    // This is less intuitive. As far as I understand, the identifier was declared;
    // it just failed to resolve fully. Even the type should be known.
    const copy = totally_a_declared_identifier;
}
```

In this trivial example it's not too severe, but it can be a pain point when there is some distance between the attempted declaration and the access, or when `S` is a large struct with one (buried) failed field. My reaction to that error message is usually "I don't understand why it says that, I definitely declared that variable" and it takes a moment to see the problem.

The error message can stays the same, but adding a note pointing to the original identifier which failed to analyze would be helpful for quickly fixing the problem:

```
    const totally_a_declared_identifier = S{ .value = bogus };
          ^
/path/to/src.zig:8:11: note: failed declaration of `totally_a_declared_identifier` here
```

As for implementation, maybe var decls which fail to fully analyze could be cached (for the duration of the current scope) and then looked up in future "undeclared identifier" errors.

An alternative would be to replace `undeclared identifier` with a more accurate explanation such as `identifier which could not be resolved` or something.

**Comment by andrewrk** [comment]
Huh, that's weird, I thought I solved this with  #111. Thanks for the report.

**Comment by Vexu** [comment]
Doesn't happen in master.
